ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -645,21 +651,19 @@ VALUE Message_decode(VALUE klass, VALUE data) {
 
   const upb_pbdecodermethod* method = msgdef_decodermethod(desc);
   const upb_handlers* h = upb_pbdecodermethod_desthandlers(method);
-  upb_pbdecoder decoder;
-  upb_sink sink;
-  upb_status status = UPB_STATUS_INIT;
+  upb_env e;
+  upb_env_init(&e);
```

Comment:
I ran the following benchmark against old and new:

    #!/usr/bin/ruby                                                                                                                                                                                                      
                                                                                                                                                                                                                         
    require 'google/protobuf'                                                                                                                                                                                            
                                                                                                                                                                                                                         
    p = Google::Protobuf::DescriptorPool.generated_pool                                                                                                                                                                  
    p.build do                                                                                                                                                                                                           
      add_message "TestMessage" do                                                                                                                                                                                       
        repeated :foo, :int32, 1                                                                                                                                                                                         
      end                                                                                                                                                                                                                
    end                                                                                                                                                                                                                  
                                                                                                                                                                                                                         
    TestMessage = p.lookup("TestMessage").msgclass                                                                                                                                                                       
                                                                                                                                                                                                                         
    testmsg = TestMessage.new(:foo => (1..100).to_a)                                                                                                                                                                     
    data = TestMessage.encode(testmsg)                                                                                                                                                                                   
                                                                                                                                                                                                                         
    trials = []                                                                                                                                                                                                          
    10.times do                                                                                                                                                                                                          
      t1 = Time.now.to_f                                                                                                                                                                                                 
      100_000.times do                                                                                                                                                                                                   
        m = TestMessage.decode(data)                                                                                                                                                                                     
      end                                                                                                                                                                                                                
      t2 = Time.now.to_f                                                                                                                                                                                                 
      trials << (t2 - t1)                                                                                                                                                                                                
    end                                                                                                                                                                                                                  
                                                                                                                                                                                                                         
    puts "trials: #{trials}"                                                                                                                                                                                             
    puts "avg: #{trials.reduce(:+)/trials.length}" 

And found average time-per-trial to be ~0.51s for old and ~0.53s for new. This is with small (200 byte) and simple (int32 only) test messages. (As an aside, this comes to ~40 MB/s). So there is a ~4% impact. IMHO that's enough to "do the right thing" here -- I'll use `seededalloc` as you suggest.